### PR TITLE
JGoogleEmbedMaps::addMarker with event(s)

### PR DESCRIPTION
**Pull request by itbra**
I found that when adding a marker to the map it is not considered to bind events to it. Isn't it a tyical scenario to click on a marker to show a balloon with short info or a link to more details? 

The [Google Maps API](https://developers.google.com/maps/documentation/javascript/reference) presents as an example to create the marker and immediately create the desired event which expects to receive the marker created as reference to bind to. I extended those functions responsible for marker creation by adding them another argument for the events _([find a PR for map events here](https://github.com/joomla/joomla-cms/pull/6472))_. Also, the function creating the script was extended to evaluate the currently set markers not only for their options but also for events to bind.

I tested this functionality by taking a view template and paste the following:

```
// Target element
echo '<h1>JGoogleEmbedMaps Test</h1>';
echo '<div class="row">';
echo    '<div class="col col-xs-12">';
echo        '<div id="map" style="height:500px; background:gainsboro; border:gray">LOADING ...</div>';
echo    '</div>';
echo    '<div class="col col-xs-12">';
echo        '<h5>Monitor</h5>';
echo        '<p id="monitor" style="padding:1em 0.5em; border:1px solid gainsboro; border-radius:3px"></p>';
echo    '</div>';
echo '</div>';

// Config
$options = new JRegistry(array(
    'key'         => '',
    'mapid'       => 'map',
    'mapclass'    => 'mapclass'
));

// Get instance
$google = new JGoogle($options);
$map = $google->embed('Maps');

// Define load type
$map->setAutoload('jquery');

$map->setMapType('HYBRID');
$map->setZoom('3');

// Marker configuration
$markerOptions = array(
    'clickable' => true,
    'draggable' => false,
    'opacity'   => 0.75,
    'optimized' => true,
    'visible'   => true
);

$map->setCenter(array('51.165691','10.451526'), 'Germany', $markerOptions,
array(
    'click' => 'function(e) {
        console.log("event: ", e);
        map.setCenter(this.getPosition());
        document.getElementById("monitor").innerHTML = "Clicked:&nbsp;" + this.title;
    }'
));

// Add markers
$map->addMarker(array('46.227638','2.213749'),  'France');  
$map->addMarker(array('60.472023','8.468946'),  'Norway', $markerOptions);
$map->addMarker(array('51.919437','19.145136'), 'Poland', $markerOptions, array());
$map->addMarker(array('41.871941','12.567380'), 'Italy',  $markerOptions, array(
    'click' => 'function(e) {
        console.log("event: ", e);
        map.setCenter(this.getPosition());
        document.getElementById("monitor").innerHTML = "Clicked:&nbsp;" + this.title;
    }'
));

// Prepare callback
$callback =
<<<ENDSCRIPT
window.gm = map;   // Will execute after initialisation to add reference to the created map as a global.
(function($, gm, element) {
    $('#monitor').append('<em>Callback executed</em>')
})(jQuery.noConflict(), window.google || {}, document.getElementById({$options->get('mapid')}));
ENDSCRIPT;

// Register custom script(s)
$map->setAdditionalJavascript($callback);

$map->echoHeader();
$map->echoBody();
```

So far this works. I wonder why this hasn't been considered yet. What do you think about this?

### CODE REVIEW

**Comment by itbra** [comment]
The tests appear to fail although the code works properly. Can somebody please tell me what causes the test to fail? Is there probably a change required in the test class - mabe must a test for the extra function arguments with `setCenter` and `addMarker`? Mabe the tests don't consider these arguments and cause the test to fail for that reason? 
I have no idea, but this message from travis *Failed asserting that two arrays are equal.* makes me to assume that.

**Comment by Hackwar** [comment]
The tests fail because you have a PHP syntax error in your changed test file on line 360. You are doing a ```"somethin" => "value"``` outside of an array scope. Please fix that.

**Comment by itbra** [comment]
Thanks for pointing me to the issue. Finally its passing all tests. I'll add one more test case with a non-empty events-array.

**Comment by itbra** [comment]
There we go. Extra test for empty or non-empty events array added and passed.

**Comment by designbengel** [comment]
Sorry i couldn´t test successfully. By using your code above with the patch nothing happens by click on the markers.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6381">issues.joomla.org/joomla-cms/6381</a>.</sub>

**Comment by itbra** [comment]
Of course it works. I tested it right this second via copy+paste the above example 1:1. You have to apply the patch to JGoogleEmbedMaps first so that the events are registered on marker creation.

**Comment by designbengel** [comment]
Oh sorry, don´t know how to apply something to JGoogleEmbedMaps. I think i tested it wrong.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6381">issues.joomla.org/joomla-cms/6381</a>.</sub>

**Comment by itbra** [comment]
After integrating the event binding to map markers and fixing the corresponding unit tests i'm no going to extend this PR to allow for binding events to the map itself. I will also try to adapt from the existing unit tests to create additional tests for the newly added maps methods.

File: libraries/joomla/google/embed/maps.php [hunk 1]
```diff
@@ -565,18 +587,33 @@ public function getHeader()
...
+				$options = $marker['options'];
+
+				$setup .= 'marker = new google.maps.Marker({';
+				$setup .= "position: new google.maps.LatLng({$loc[0]},{$loc[1]}),";
+				$setup .= 'map: map,';
+				$setup .= "title:'{$title}',";
+				$setup .= substr(json_encode($options), 1, -1);
+				$setup .= '});';
+
+				if (isset($marker['events']))
```
Comment (bembelimen) [inline libraries/joomla/google/embed/maps.php]: You should test here, if it's an array:

```php
if (isset($marker['events']) && is_array($marker['events']))
```

File: tests/unit/suites/libraries/joomla/google/JGoogleEmbedMapsTest.php [hunk 2]
```diff
@@ -308,12 +308,12 @@ public function testSetCenter()
 	{
 		$this->http->expects($this->exactly(5))->method('get')->will($this->returnCallback('mapsGeocodeCallback'));
 
-		$reference[] = array('loc' => array(37, -122), 'title' => '37, -122', 'options' => array());
+		$reference[] = array('loc' => array(37, -122), 'title' => '37, -122', 'options' => array(), 'events' => array());
```
Comment (bembelimen) [inline tests/unit/suites/libraries/joomla/google/JGoogleEmbedMapsTest.php]: Where is the sense in testing the empty array as parameter? That's the default anyways. It would make more sense to test invalid values etc.

File: libraries/joomla/google/embed/maps.php [hunk 3]
```diff
@@ -565,18 +587,33 @@ public function getHeader()
 		$setup .= '};';
 		$setup .= "var map = new google.maps.Map(document.getElementById('{$id}'), mapOptions);";
 
-		foreach ($this->listMarkers() as $marker)
+		$markers = $this->listMarkers();
+
+		if (isset($markers) && count($markers))
```
Comment (bembelimen) [inline libraries/joomla/google/embed/maps.php]: Could be:
```php
if (!empty($markers))
```
or better test if it is an array

Comment (itbra) [inline libraries/joomla/google/embed/maps.php]: I also already considered whether to test for `$markers` being an array, but since the method `listMarkers()` returns an array the check for that appears to be unnecessary. But i wasn't sure and wanted to wait for other opinions. Then, does it make a difference to call the method `empty()` which will itself apply a count on the received array or to apply the count directly? I think no. So i see no reason for a change.
Actually i find the current version of line 592 is unnecessary at all for serveral reasons:

a) The original version does not apply a check at all. It just starts the iteration with `foreach ($this->listMarkers() as $marker)` which is kinda statement to be sure the returned value is of type array but not considering whether it is empty.
b) The check for being not empty consumes extra time and resource and i guess this is more than just starting the loop to detect the returned array is empty and finish the loop.

So i consider to drop this check at all. Please correct me if i oversee something.

File: libraries/joomla/google/embed/maps.php (hunk 1 above)
Comment (itbra) [inline libraries/joomla/google/embed/maps.php]: Agree. I will update this line.

**Comment by Erftralle** [comment]
> What do you think about this?

I find it a usefull addition, thanks!

@test: Works fine for me.

File: libraries/joomla/google/embed/maps.php (hunk 3 above)
Comment (bembelimen) [inline libraries/joomla/google/embed/maps.php]: Nope, you're correct, my concern was not your code, but that someone could make something like:
```php
$map->setOptions('markers', 'foo');
```
and it would break. I think, the "listMarkers" method itself should test, if we have an array:
```php
 return is_array($this->getOption('markers')) ? $this->getOption('markers') : array();
```

Comment (itbra) [inline libraries/joomla/google/embed/maps.php]: This was in fact also my concern and the reason why i intended to write '$markers = (array) $this->listMarkers();' (same with listEvents()). Basically i am very much focused on data being and staying consistent and thus on type checks as well as content checks. But i also thought that when this class was written along with its test the author(s) might already have considered the return value to always be an array.
What do you think? Should the corresponding methods get updated as well? How to argue this well enough to get enough agreement for a merge?
Or shall i add my mentioned cast to where i fetch the lists? This was actually as not a good idea because there are more places where these lists are queried for and it might be no good idea to spread type casts all around.

**Comment by itbra** [comment]
@Erftralle thanks for testing and your feedback. Are there maybe more testers willing to cpy and paste the above example and report back?

**Comment by Rivaner** [comment]
I have tested this item :white_check_mark: successfully on 7bb7f93d46509ad305e96df1d4e3140290470205<br /><br />It works.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6381">issues.joomla.org/joomla-cms/6381</a>.</sub>

**Comment by DavidTrenkle** [comment]
I have tested this item :white_check_mark: successfully on 7bb7f93d46509ad305e96df1d4e3140290470205<br /><br />@test works with the example code<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6381">issues.joomla.org/joomla-cms/6381</a>.</sub>

**Comment by KlausKrippeit** [comment]
I have tested this item :white_check_mark: successfully on 7bb7f93d46509ad305e96df1d4e3140290470205<br /><br />@test successful<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6381">issues.joomla.org/joomla-cms/6381</a>.</sub>

**Comment by KlausKrippeit** [comment]
I have tested this item :white_check_mark: successfully on 7bb7f93d46509ad305e96df1d4e3140290470205<br /><br />@test successful<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6381">issues.joomla.org/joomla-cms/6381</a>.</sub>

**Comment by zero-24** [comment]
RTC. Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6381">issues.joomla.org/joomla-cms/6381</a>.</sub>

**Comment by roland-d** [comment]
Thanks everybody. Merged into 3.5-dev with commit  269b781
